File: src/matrix_multiplication.c
/**
 * Author: Jack Robbins
 * This file contains a simple C program that compares the execution time of multiplying a 4 x 4 matrix
 * with its inverse using the traditional matrix multiplication algorithm and Strassen's algorithm
 */

#include <stdio.h>
#include <stdlib.h>

void print_matrix(double** matrix, int N){
	for(int i = 0; i < N; i++){
		for(int j = 0; j < N; j++){
			printf("%.2f ", matrix[i][j]);
		}
	printf("\n");
	}
}

double** matrix_mult_classic(double** matrix_a, double** matrix_b){

}


double** matrix_mult_strassen(double** matrix_a, double** matrix_b){

}


int main(int argc, char** argv){
	//The first command line argument will be the dimension of the matrix
	
	const int N = atoi(argv[1]);

	//If we don't see the right amount of arguments, hard exit
	if(argc != N * N + 2){
		printf("Incorrect number of inputs provided for an %d X %d matrix\n", N, N);
		printf("%d\n", argc);
		exit(1);
	}

	//Declare and allocate our N x N matrix
	double** matrix = (double**)malloc(sizeof(double*) * N);
	double** inverse = (double**)malloc(sizeof(double*) * N);

	for(int i = 0; i < N; i++){
		matrix[i] = (double*)malloc(sizeof(double) * N);
		inverse[i] = (double*)malloc(sizeof(double) * N);
	}

	//Advance the pointer by 2
	argv += 2;

	//For running through argv
	int index = 0;

	//Grab all of our inputs
	for(int i = 0; i < N; i++){
		for(int j = 0; j < N; j++){
			matrix[i][j] = atof(argv[index]);	
		 	index++;
		}
	}
	
	printf("Inputted Matrix:\n");
	print_matrix(matrix, N);

	print_matrix(inverse, N);

}

File: src/matrix_multiplication.py
"""
Author: Jack Robbins
This python file will multiply a matrix with its inverse using the regular method
and Strassen's algorithm to compare the performance of the two methods
"""

import numpy as np
import time


def matrix_mult_classic(a, b):
    result = np.array([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
    start_time = time.time()


    end_time = time.time()
    

def matrix_mult_strassen(a, b):
    result = np.array([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
    start_time = time.time()


    end_time = time.time() 

"""
Main method for running/printing
"""
def main():
    print("Testing matrix multiplication execution time.")
    print("Input Matrix:")
    #Matrix given in prompt
    matrix = np.array([[1, 3, 3, 6], [4, 2, 8, 2],[3, 3, 4, 5],[2, 6, 3, 1]])
    print(matrix)
    print("\n")

    # Calculate inverse
    inverse = np.linalg.inv(matrix)
    print("Inverse of input:")
    print(inverse)
    print("\n")


if __name__ == '__main__':
    main()
